In [26]:
import pandas as pd

In [27]:
df = pd.read_csv('Airbnb_Open_Data.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

/var/folders/vn/g54cndv11_98hbh4s_vfvw4h0000gp/T/ipykernel_14796/857120404.py:1: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Airbnb_Open_Data.csv')


In [28]:
df.head(5)

,id,NAME,host id,host_identity_verified,host name,neighbourhood group,neighbourhood,lat,long,country,...,service fee,minimum nights,number of reviews,last review,reviews per month,review rate number,calculated host listings count,availability 365,house_rules,license
0,1001254,Clean & quiet apt home by the park,80014485718,unconfirmed,Madaline,Brooklyn,Kensington,40.64749,-73.97237,United States,...,$193,10.0,9.0,10/19/2021,0.21,4.0,6.0,286.0,Clean up and treat the home the way you'd like...,NaN
1,1002102,Skylit Midtown Castle,52335172823,verified,Jenna,Manhattan,Midtown,40.75362,-73.98377,United States,...,$28,30.0,45.0,5/21/2022,0.38,4.0,2.0,228.0,Pet friendly but please confirm with me if the...,NaN
2,1002403,THE VILLAGE OF HARLEM....NEW YORK !,78829239556,NaN,Elise,Manhattan,Harlem,40.80902,-73.94190,United States,...,$124,3.0,0.0,NaN,NaN,5.0,1.0,352.0,"I encourage you to use my kitchen, cooking and...",NaN
3,1002755,NaN,85098326012,unconfirmed,Garry,Brooklyn,Clinton Hill,40.68514,-73.95976,United States,...,$74,30.0,270.0,7/5/2019,4.64,4.0,1.0,322.0,NaN,NaN
4,1003689,Entire Apt: Spacious Studio/Loft by central park,92037596077,verified,Lyndon,Manhattan,East Harlem,40.79851,-73.94399,United States,...,$41,10.0,9.0,11/19/2018,0.10,3.0,1.0,289.0,"Please no smoking in the house, porch or on th...",NaN


**Data Description**

| #  | Column                          | Non-Null Count | Dtype    |
|----|----------------------------------|----------------|----------|
| 0  | id                               | 102599         | int64    |
| 1  | NAME                             | 102349         | object   |
| 2  | host id                          | 102599         | int64    |
| 3  | host_identity_verified           | 102310         | object   |
| 4  | host name                        | 102193         | object   |
| 5  | neighbourhood group              | 102570         | object   |
| 6  | neighbourhood                    | 102583         | object   |
| 7  | lat                              | 102591         | float64  |
| 8  | long                             | 102591         | float64  |
| 9  | country                          | 102067         | object   |
| 10 | country code                     | 102468         | object   |
| 11 | instant_bookable                 | 102494         | object   |
| 12 | cancellation_policy              | 102523         | object   |
| 13 | room type                        | 102599         | object   |
| 14 | Construction year                | 102385         | float64  |
| 15 | price                            | 102352         | object   |
| 16 | service fee                      | 102326         | object   |
| 17 | minimum nights                   | 102190         | float64  |
| 18 | number of reviews                | 102416         | float64  |
| 19 | last review                      | 86706          | object   |
| 20 | reviews per month                | 86720          | float64  |
| 21 | review rate number               | 102273         | float64  |
| 22 | calculated host listings count   | 102280         | float64  |
| 23 | availability 365                 | 102151         | float64  |
| 24 | house_rules                      | 50468          | object   |
| 25 | license                          | 2              | object   |


**Data Preprocesisng**


In [29]:
# df.drop(columns = [license], inplace= True)
list(df.columns)
df.drop(columns = ['license'], inplace= True)

In [30]:
df.head()

,id,NAME,host id,host_identity_verified,host name,neighbourhood group,neighbourhood,lat,long,country,...,price,service fee,minimum nights,number of reviews,last review,reviews per month,review rate number,calculated host listings count,availability 365,house_rules
0,1001254,Clean & quiet apt home by the park,80014485718,unconfirmed,Madaline,Brooklyn,Kensington,40.64749,-73.97237,United States,...,$966,$193,10.0,9.0,10/19/2021,0.21,4.0,6.0,286.0,Clean up and treat the home the way you'd like...
1,1002102,Skylit Midtown Castle,52335172823,verified,Jenna,Manhattan,Midtown,40.75362,-73.98377,United States,...,$142,$28,30.0,45.0,5/21/2022,0.38,4.0,2.0,228.0,Pet friendly but please confirm with me if the...
2,1002403,THE VILLAGE OF HARLEM....NEW YORK !,78829239556,NaN,Elise,Manhattan,Harlem,40.80902,-73.94190,United States,...,$620,$124,3.0,0.0,NaN,NaN,5.0,1.0,352.0,"I encourage you to use my kitchen, cooking and..."
3,1002755,NaN,85098326012,unconfirmed,Garry,Brooklyn,Clinton Hill,40.68514,-73.95976,United States,...,$368,$74,30.0,270.0,7/5/2019,4.64,4.0,1.0,322.0,NaN
4,1003689,Entire Apt: Spacious Studio/Loft by central park,92037596077,verified,Lyndon,Manhattan,East Harlem,40.79851,-73.94399,United States,...,$204,$41,10.0,9.0,11/19/2018,0.10,3.0,1.0,289.0,"Please no smoking in the house, porch or on th..."


In [31]:
#Display the country code 
country_code = df['country code']
print(country_code)

0         US
1         US
2         US
3         US
4         US
          ..
102594    US
102595    US
102596    US
102597    US
102598    US
Name: country code, Length: 102599, dtype: object


In [32]:
#Chuan hoa ten cot (Column name standardization, lower case and replace space with underscore)
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
)

df.columns

Index(['id', 'name', 'host_id', 'host_identity_verified', 'host_name',
       'neighbourhood_group', 'neighbourhood', 'lat', 'long', 'country',
       'country_code', 'instant_bookable', 'cancellation_policy', 'room_type',
       'construction_year', 'price', 'service_fee', 'minimum_nights',
       'number_of_reviews', 'last_review', 'reviews_per_month',
       'review_rate_number', 'calculated_host_listings_count',
       'availability_365', 'house_rules'],
      dtype='object')

In [34]:
hosts_df = df[[
    "host_id",
    "host_name",
    "host_identity_verified",
    "calculated_host_listings_count"
]].drop_duplicates()
hosts_df.to_csv("db-airbnb/hosts.csv", index=False)
hosts_df.head()

,host_id,host_name,host_identity_verified,calculated_host_listings_count
0,80014485718,Madaline,unconfirmed,6.0
1,52335172823,Jenna,verified,2.0
2,78829239556,Elise,NaN,1.0
3,85098326012,Garry,unconfirmed,1.0
4,92037596077,Lyndon,verified,1.0


In [35]:
locations_df = df[[
    "neighbourhood_group",
    "neighbourhood",
    "lat",
    "long",
    "country",
    "country_code"
]].drop_duplicates().reset_index(drop=True)
locations_df.insert(0, "location_id", locations_df.index + 1)
locations_df.to_csv("db-airbnb/locations.csv", index=False)
locations_df.head()

,location_id,neighbourhood_group,neighbourhood,lat,long,country,country_code
0,1,Brooklyn,Kensington,40.64749,-73.97237,United States,US
1,2,Manhattan,Midtown,40.75362,-73.98377,United States,US
2,3,Manhattan,Harlem,40.80902,-73.94190,United States,US
3,4,Brooklyn,Clinton Hill,40.68514,-73.95976,United States,US
4,5,Manhattan,East Harlem,40.79851,-73.94399,United States,US


In [36]:
df = df.merge(
    locations_df,
    on=["neighbourhood_group", "neighbourhood", "lat", "long", "country", "country_code"],
    how="left"
)
df.head()

,id,name,host_id,host_identity_verified,host_name,neighbourhood_group,neighbourhood,lat,long,country,...,service_fee,minimum_nights,number_of_reviews,last_review,reviews_per_month,review_rate_number,calculated_host_listings_count,availability_365,house_rules,location_id
0,1001254,Clean & quiet apt home by the park,80014485718,unconfirmed,Madaline,Brooklyn,Kensington,40.64749,-73.97237,United States,...,$193,10.0,9.0,10/19/2021,0.21,4.0,6.0,286.0,Clean up and treat the home the way you'd like...,1
1,1002102,Skylit Midtown Castle,52335172823,verified,Jenna,Manhattan,Midtown,40.75362,-73.98377,United States,...,$28,30.0,45.0,5/21/2022,0.38,4.0,2.0,228.0,Pet friendly but please confirm with me if the...,2
2,1002403,THE VILLAGE OF HARLEM....NEW YORK !,78829239556,NaN,Elise,Manhattan,Harlem,40.80902,-73.94190,United States,...,$124,3.0,0.0,NaN,NaN,5.0,1.0,352.0,"I encourage you to use my kitchen, cooking and...",3
3,1002755,NaN,85098326012,unconfirmed,Garry,Brooklyn,Clinton Hill,40.68514,-73.95976,United States,...,$74,30.0,270.0,7/5/2019,4.64,4.0,1.0,322.0,NaN,4
4,1003689,Entire Apt: Spacious Studio/Loft by central park,92037596077,verified,Lyndon,Manhattan,East Harlem,40.79851,-73.94399,United States,...,$41,10.0,9.0,11/19/2018,0.10,3.0,1.0,289.0,"Please no smoking in the house, porch or on th...",5


In [94]:
rooms_df = df[[
    "id",
    "name",
    "host_id",
    "location_id",
    "room_type",
    "construction_year",
    "minimum_nights",
    "availability_365",
    "house_rules",
    "cancellation_policy",
    "instant_bookable"
]].drop_duplicates()
rooms_df = rooms_df.rename(columns={"id": "room_id"})
rooms_df["construction_year"] = (
    rooms_df["construction_year"]
    .astype(float)
    .round()
    .astype("Int64")
)
rooms_df["minimum_nights"] = (
    rooms_df["minimum_nights"]
    .astype(float)
    .round()
    .astype("Int64")
)
rooms_df["availability_365"] = (
    rooms_df["availability_365"]
    .astype(float)
    .round()
    .astype("Int64")
)

rooms_df.to_csv("db-airbnb/rooms.csv", index=False)
rooms_df.head()


,room_id,name,host_id,location_id,room_type,construction_year,minimum_nights,availability_365,house_rules,cancellation_policy,instant_bookable
0,1001254,Clean & quiet apt home by the park,80014485718,1,Private room,2020,10,286,Clean up and treat the home the way you'd like...,strict,False
1,1002102,Skylit Midtown Castle,52335172823,2,Entire home/apt,2007,30,228,Pet friendly but please confirm with me if the...,moderate,False
2,1002403,THE VILLAGE OF HARLEM....NEW YORK !,78829239556,3,Private room,2005,3,352,"I encourage you to use my kitchen, cooking and...",flexible,True
3,1002755,NaN,85098326012,4,Entire home/apt,2005,30,322,NaN,moderate,True
4,1003689,Entire Apt: Spacious Studio/Loft by central park,92037596077,5,Entire home/apt,2009,10,289,"Please no smoking in the house, porch or on th...",moderate,False


In [ ]:
# pricing_df = df[[
#     "id",
#     "price",
#     "service_fee"
# ]].drop_duplicates()
# pricing_df = pricing_df.rename(columns={"id": "room_id"})
# pricing_df.insert(0, "pricing_id", range(1, len(pricing_df) + 1))
# pricing_df["effective_from"] = pd.to_datetime("today").normalize()
# pricing_df.to_csv("db-airbnb/pricing.csv", index=False)
# pricing_df.head()


,pricing_id,room_id,price,service_fee,effective_from
0,1,1001254,$966,$193,2026-01-25
1,2,1002102,$142,$28,2026-01-25
2,3,1002403,$620,$124,2026-01-25
3,4,1002755,$368,$74,2026-01-25
4,5,1003689,$204,$41,2026-01-25


In [88]:
review_stats_df = df[[
    "id",
    "number_of_reviews",
    "last_review",
    "reviews_per_month",
    "review_rate_number"
]].drop_duplicates()

review_stats_df = review_stats_df.rename(columns={"id": "room_id"})

# ===== CLEAN DTYPE =====

# room_id: int
review_stats_df["room_id"] = review_stats_df["room_id"].astype("Int64")

# number_of_reviews: float -> int
review_stats_df["number_of_reviews"] = (
    review_stats_df["number_of_reviews"]
    .astype(float)
    .round()
    .astype("Int64")
)

# last_review: object -> date
review_stats_df["last_review"] = (
    pd.to_datetime(review_stats_df["last_review"], errors="coerce")
    .dt.date
)

# reviews_per_month: float -> 2 decimals
review_stats_df["reviews_per_month"] = (
    review_stats_df["reviews_per_month"]
    .astype(float)
    .round(2)
)

# review_rate_number: float -> 2 decimals
review_stats_df["review_rate_number"] = (
    review_stats_df["review_rate_number"]
    .astype(float)
    .round(2)
)

# ===== EXPORT =====
review_stats_df.to_csv("db-airbnb/review_stats.csv", index=False)

review_stats_df.head()


,room_id,number_of_reviews,last_review,reviews_per_month,review_rate_number
0,1001254,9,2021-10-19,0.21,4.0
1,1002102,45,2022-05-21,0.38,4.0
2,1002403,0,NaT,NaN,5.0
3,1002755,270,2019-07-05,4.64,4.0
4,1003689,9,2018-11-19,0.10,3.0


In [90]:
review_stats_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 102058 entries, 0 to 102057
Data columns (total 5 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   room_id             102058 non-null  Int64  
 1   number_of_reviews   101875 non-null  Int64  
 2   last_review         86226 non-null   object 
 3   reviews_per_month   86240 non-null   float64
 4   review_rate_number  101739 non-null  float64
dtypes: Int64(2), float64(2), object(1)
memory usage: 4.9+ MB


In [56]:
pricing_df = df[[
    "id",
    "price",
    "service_fee"
]].drop_duplicates()
pricing_df = pricing_df.rename(columns={"id": "room_id"})
pricing_df.insert(0, "pricing_id", range(1, len(pricing_df) + 1))
pricing_df["effective_from"] = pd.to_datetime("today").normalize()
pricing_df["price"] = (
    pricing_df["price"]
      .str.replace("$", "", regex=False)
      .str.replace(",", "", regex=False)
)
pricing_df["service_fee"] = (
    pricing_df["service_fee"]
      .str.replace("$", "", regex=False)
      .str.replace(",", "", regex=False)
)
pricing_df.to_csv("db-airbnb/pricing.csv", index=False)

In [57]:
price_df = pd.read_csv('db-airbnb/pricing.csv')
price_df.head()

,pricing_id,room_id,price,service_fee,effective_from
0,1,1001254,966.0,193.0,2026-01-25
1,2,1002102,142.0,28.0,2026-01-25
2,3,1002403,620.0,124.0,2026-01-25
3,4,1002755,368.0,74.0,2026-01-25
4,5,1003689,204.0,41.0,2026-01-25


In [82]:
hosts_df = df[[
    "host_id",
    "host_name",
    "host_identity_verified",
    "calculated_host_listings_count"
]].drop_duplicates()
# hosts_df.to_csv("db-airbnb/hosts.csv", index=False)
# # Fix boolean
hosts_df["host_identity_verified"] = hosts_df["host_identity_verified"].str.lower().map({
    "verified": True,
    "unconfirmed": False,
    "true": True,
    "false": False
})
# # Fix int column
hosts_df["calculated_host_listings_count"] = (
    hosts_df["calculated_host_listings_count"]
    .fillna(0)
    .astype(int)
)
hosts_df.to_csv("db-airbnb/hosts.csv", index=False)

In [83]:
hosts_df.head()

,host_id,host_name,host_identity_verified,calculated_host_listings_count
0,80014485718,Madaline,False,6
1,52335172823,Jenna,True,2
2,78829239556,Elise,NaN,1
3,85098326012,Garry,False,1
4,92037596077,Lyndon,True,1


In [84]:
# count duplicate host_id entries
hosts_dup = hosts_df[hosts_df.duplicated(subset=["host_id"], keep=False)]
hosts_dup.sort_values("host_id").to_csv("hosts_duplicates.csv", index=False)